### ISSUE TITLE
Possible factorial error.

### ISSUE THREAD

**Issue by a35joshi** [fasseg/exp4j#75]
Hi.
I am not sure about this being a bug or not but I will put forward some of my findings.
So I defined the factorial method as given on [(http://www.objecthunter.net/exp4j/)]. 
I get an exception for 3!-2! but not for 3-2!(I get 1, as expected). But righfully, (3!)-(2!) gives me the correct answer(i.e, 4). 
What can be done in order for 3!-2! to work without having to enter brackets?

**Comment by dktcoding** [comment]
Yes, this actually related to #64 see pull request #74, I just added a test for this and it seems to be resolved in that branch (https://travis-ci.org/dktcoding/exp4j/builds/182453474)

**Comment by a35joshi** [comment]
case '!':
        if (numArguments != 1) {
        return BUILTIN[INDEX_FACTORIAL];
        }
        default:
        return null;


This is the proposed fix correct? In this, would the numArguments be number of elements in the string to which I add operators and numbers?


________________________________

**Comment by dktcoding** [comment]
That's the special casing, the way it works is:
1. The `!` is declared to have only one argument (as it should)
1. Since we need `!` to be always on the right we treated as a binary operator with a `null` second argument (like `num!null`)
1. The tokenizer doesn't actually count the number of available tokens, it uses some clever assumptions to try and figure out the [argument count](https://github.com/fasseg/exp4j/blob/master/src/main/java/net/objecthunter/exp4j/tokenizer/Tokenizer.java#L207-L231).
1. Based on (3) we could trick it to believe that it should have two arguments when there's only one is available, that way the expression is always valid (because it has only one argument), but the operator is always on the right.

The problem is that this can't be done using an external `Operator` (at least I was unable to figure out how), only using a _built-in_ so it's possible to _special case it_

That code you posted isn't from the original `exp4j` it's from a fork I created, the proposed fix is detailed in PR #74

**Comment by fasseg** [comment]
Hey Guys, 
I do have January off and will invest some time to get exp4j up to date. Unfortunately I do not have enough time to do this properly earlier...

**Comment by fasseg** [comment]
Hey Guys, 

I added the tests from @dktcoding and it does produce the correct value. Am I missing something there or has this bug mysteriously disappeared?
https://github.com/fasseg/exp4j/commit/e977f199cc1222cd5232582a03a3742324710b9a
